### PR TITLE
Fix detection of integral dtypes

### PR DESCRIPTION
**Pull request by dhirschfeld**
**TL;DR;**
`if ty in datashape.integral` doesn't detect `?decimal[precision=1, scale=0]` as integral but `ty.to_numpy_dtype()` returns `np.int32` which then blows up on nulls (None)

**Also Fixes:**
- #556
- #597

### CODE REVIEW

**Comment by dhirschfeld** [comment]
**Repro**:
```python
import blaze as bz
from numpy.random import randn
from odo import odo
import pandas as pd
import sqlalchemy as sa


engine = sa.create_engine('sqlite://')
metadata = sa.MetaData(bind=engine)


T = sa.Table(
    'Demo',
    metadata,
    sa.Column('pkid', sa.Integer, primary_key=True),
    sa.Column('value', sa.DECIMAL(precision=1, scale=0, asdecimal=False), nullable=True),
)

metadata.create_all()

data = pd.DataFrame(list(zip(range(10), randn(10) > 0)), columns=['pkid','value'])
data.loc[5, 'value'] = None
records = data.to_dict(orient='records')

with engine.connect() as conn:
    conn.execute(T.insert(), records)
```
```python
In [2]: T.select().execute().fetchall()
Out[2]: 
[(0, 0),
 (1, 1),
 (2, 1),
 (3, 0),
 (4, 0),
 (5, None),
 (6, 0),
 (7, 0),
 (8, 1),
 (9, 0)]

In [4]: bz.discover(T)
Out[4]: dshape("var * {pkid: int32, value: ?decimal[precision=1, scale=0]}")

In [5]: odo(T, pd.DataFrame)
Traceback (most recent call last):

  File "<ipython-input-5-4015a811a204>", line 1, in <module>
    odo(T, pd.DataFrame)

  File "C:\Miniconda3\lib\site-packages\odo\odo.py", line 91, in odo
    return into(target, source, **kwargs)

  File "C:\Miniconda3\lib\site-packages\multipledispatch\dispatcher.py", line 164, in __call__
    return func(*args, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\into.py", line 43, in wrapped
    return f(*args, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\into.py", line 53, in into_type
    return convert(a, b, dshape=dshape, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\core.py", line 83, in __call__
    return _transform(self.graph, *args, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\core.py", line 106, in _transform
    x = f(x, excluded_edges=excluded_edges, **kwargs)

  File "C:\Miniconda3\lib\site-packages\odo\backends\sql.py", line 757, in select_or_selectable_to_frame
    dtype=[(str(c), dtypes[c]) for c in columns]))

TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

**Comment by dhirschfeld** [comment]
```
<snip>
odo/backends/tests/test_sql.py::test_integer_detection PASSED
<snip>
======= 457 passed, 45 skipped, 8 xfailed, 5 warnings in 27.74 seconds ========
```

**Comment by dhirschfeld** [comment]
@llllllllll, @ehebert, @postelrich - it would be good if someone could hit the merge button on this one as it fixes a number of issues, including compatibility with the latest pandas. As mentioned above, the travis failure seems unrelated.

**Comment by necaris** [comment]
LGTM

**Comment by dhirschfeld** [comment]
Could this PR please be merged - it fixes a number of bugs which are affecting my use of odo in a production system.

@llllllllll - I understand that there's no full-time maintainer for odo but if no one else has time to review and merge bug-fix PRs perhaps I could be given commit rights myself to do so?

**Comment by ehebert** [comment]
@dhirschfeld to help expedite a merge I can offer to review your work soon. However, I currently also do not have commit rights.

**Comment by dhirschfeld** [comment]
@llllllllll - any chance you could merge this? Happy to make any suggested changes...

File: odo/backends/sql.py [hunk 1]
```diff
@@ -794,13 +794,14 @@ def select_or_selectable_to_frame(el, bind=None, dshape=None, **kwargs):
...
+            try:
+                dtypes[field] = ty.to_numpy_dtype()
+            except TypeError:
+                dtypes[field] = np.dtype(object)
             else:
-                try:
-                    dtypes[field] = ty.to_numpy_dtype()
-                except TypeError:
-                    dtypes[field] = np.dtype(object)
+                if np.issubdtype(dtypes[field], np.integer):
```
Comment (postelrich) [inline odo/backends/sql.py]: Does this else-if branch leave the possibility of no type being converted for that field?

Comment (dhirschfeld) [reply odo/backends/sql.py]: It's a try/except/else branch so the `if np.issubdtype(dtypes[field], np.integer)` block runs only if the try has already *succeeded*:
```python
try:
    dtypes[field] = ty.to_numpy_dtype()
except TypeError:
    dtypes[field] = np.dtype(object)
else:
    # If `ty.to_numpy_dtype()` results in an integer type override `dtypes[field]` to be float64
    # We do this because we already know `dtype` is an `Option` type and pandas uses NaN
    # as a null placeholder requiring that the type be floating point rather than integral
    if np.issubdtype(dtypes[field], np.integer):
        # cast nullable ints to float64 so NaN can be used for nulls
        dtypes[field] = np.float64
```

Comment (dhirschfeld) [reply odo/backends/sql.py]: 26e300e5219cbc667831b54ba55254a58fae75ba is a self-contained repro of the underlying issue

**Comment by ehebert** [comment]
@dhirschfeld I've blocked out some time today to work on merging this.

File: odo/backends/tests/test_bokeh.py [hunk 2]
```diff
@@ -15,9 +15,8 @@
 
 def test_convert_dataframe_to_cds():
     cds = convert(ColumnDataSource, df)
-    assert cds.data == {'name': ['Alice', 'Bob', 'Charlie'],
-                        'balance': [100, 200, 300]}
-
+    assert list(cds.data['name']) == ['Alice', 'Bob', 'Charlie']
```
Comment (ehebert) [inline odo/backends/tests/test_bokeh.py]: This is orthogonal to the integral dtype change, correct?
(It's ok to leave in, especially since we want to  move towards getting tests passing. Just wanted to make sure that I wasn't missing some relation to the integral dtype detection.)

Comment (dhirschfeld) [reply odo/backends/tests/test_bokeh.py]: Yes it prob should have be in a separate "Fix the tests" PR. I fixed 2 tests trying to get the CI green on this PR before I gave up. I'm happy to look into the other failures and see what I can do there to finish the job.

If I get around to it I'll make sure to open an issue first so people know I'm looking into it and we don't duplicate any work.
